File: backends/java/zappy-adapters/src/test/java/com/zappymart/adapters/persistence/CachedProductRepositoryTest.java
package com.zappymart.adapters.persistence;

import com.zappymart.application.ports.ProductRepository;
import com.zappymart.domain.catalogue.Category;
import com.zappymart.domain.catalogue.Product;
import com.zappymart.domain.shared.Money;
import org.junit.jupiter.api.Test;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;

import static org.assertj.core.api.Assertions.assertThat;

class CachedProductRepositoryTest {

    private static final Category ELECTRONICS =
            new Category("category-electronics", "Electronics", "electronics");

    private static final Product DRIVE = new Product("product-12", "WD 4TB Gaming Drive",
            "wd-4tb-gaming-drive-playstation-4", "A drive", Money.euro(11400), ELECTRONICS, 1, null);

    private final AtomicInteger readsFromTheDatabase = new AtomicInteger();

    private final AtomicInteger reducedStock = new AtomicInteger();

    private final ProductRepository behindTheCache = new ProductRepository() {
        @Override
        public List<Product> catalogue() {
            readsFromTheDatabase.incrementAndGet();
            return List.of(DRIVE);
        }

        @Override
        public Optional<Product> byId(String productId) {
            throw new AssertionError("The cache answers this one from the catalogue it remembers");
        }

        @Override
        public Optional<Product> bySlug(String slug) {
            throw new AssertionError("The cache answers this one from the catalogue it remembers");
        }

        @Override
        public List<Product> byIds(Collection<String> productIds) {
            throw new AssertionError("The cache answers this one from the catalogue it remembers");
        }

        @Override
        public void reduceStock(String productId, int quantity) {
            reducedStock.addAndGet(quantity);
        }
    };

    private final CachedProductRepository cache = new CachedProductRepository(behindTheCache);

    @Test
    void readsTheCatalogueOnceAndAnswersEveryQuestionFromIt() {
        assertThat(cache.catalogue()).containsExactly(DRIVE);
        assertThat(cache.catalogue()).containsExactly(DRIVE);
        assertThat(cache.byId("product-12")).contains(DRIVE);
        assertThat(cache.bySlug("wd-4tb-gaming-drive-playstation-4")).contains(DRIVE);
        assertThat(cache.byIds(List.of("product-12"))).containsExactly(DRIVE);
        assertThat(cache.byId("product-99")).isEmpty();

        assertThat(readsFromTheDatabase).hasValue(1);
    }

    @Test
    void readsAgainAfterAStockChange() {
        cache.catalogue();

        cache.reduceStock("product-12", 1);
        cache.catalogue();

        assertThat(reducedStock).hasValue(1);
        assertThat(readsFromTheDatabase).hasValue(2);
    }

    @Test
    void readsAgainAfterTheSeedIsLoadedAgain() {
        cache.catalogue();

        cache.forget();
        cache.catalogue();

        assertThat(readsFromTheDatabase).hasValue(2);
    }
}

File: backends/kotlin/zappy-adapters/src/test/kotlin/nl/zappymart/adapters/persistence/CachedProductRepositoryTest.kt
package nl.zappymart.adapters.persistence

import nl.zappymart.application.ports.ProductRepository
import nl.zappymart.domain.builders.aProduct
import nl.zappymart.domain.catalogue.Product
import nl.zappymart.domain.catalogue.ProductSpecification
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class CachedProductRepositoryTest {

    private class CountingProductRepository : ProductRepository {

        var pagesRead = 0
        var countsRead = 0
        var singleProductsRead = 0

        private val catalogue = listOf(aProduct().withId("product-01").named("Cotton jacket").build())

        override fun page(specification: ProductSpecification, size: Int, afterProductId: String?): List<Product> {
            pagesRead += 1
            return catalogue
        }

        override fun count(specification: ProductSpecification): Int {
            countsRead += 1
            return catalogue.size
        }

        override fun findById(productId: String): Product? {
            singleProductsRead += 1
            return catalogue.firstOrNull { product -> product.id == productId }
        }

        override fun findBySlug(slug: String): Product? {
            singleProductsRead += 1
            return catalogue.firstOrNull { product -> product.slug == slug }
        }

        override fun findAllByIds(productIds: List<String>): List<Product> {
            singleProductsRead += 1
            return catalogue.filter { product -> productIds.contains(product.id) }
        }

        override fun reduceStock(quantityPerProductId: Map<String, Int>) = Unit
    }

    private val catalogue = CountingProductRepository()

    private val cached = CachedProductRepository(catalogue)

    @Test
    fun `a page is read once and answered from the cache after that`() {
        cached.page(ProductSpecification.EVERYTHING, 5, null)
        cached.page(ProductSpecification.EVERYTHING, 5, null)
        assertThat(catalogue.pagesRead).isEqualTo(1)
    }

    @Test
    fun `a different page is a different question`() {
        cached.page(ProductSpecification.EVERYTHING, 5, null)
        cached.page(ProductSpecification.EVERYTHING, 5, "product-01")
        cached.page(ProductSpecification.of("jewellery", null, null), 5, null)
        assertThat(catalogue.pagesRead).isEqualTo(3)
    }

    @Test
    fun `a count is read once and answered from the cache after that`() {
        assertThat(cached.count(ProductSpecification.EVERYTHING)).isEqualTo(1)
        assertThat(cached.count(ProductSpecification.EVERYTHING)).isEqualTo(1)
        assertThat(catalogue.countsRead).isEqualTo(1)
    }

    @Test
    fun `reserving stock makes the cache read the catalogue again`() {
        cached.page(ProductSpecification.EVERYTHING, 5, null)
        cached.reduceStock(mapOf("product-01" to 1))
        cached.page(ProductSpecification.EVERYTHING, 5, null)
        assertThat(catalogue.pagesRead).isEqualTo(2)
    }

    @Test
    fun `one product is never answered from the cache, because its stock has to be fresh`() {
        cached.findById("product-01")
        cached.findById("product-01")
        cached.findBySlug("cotton-jacket")
        cached.findAllByIds(listOf("product-01"))
        assertThat(catalogue.singleProductsRead).isEqualTo(4)
    }
}
